### ISSUE TITLE
revamp/replace pointsinside

### ISSUE THREAD

**Issue by jefferis** [natverse/nat#353]
Still having problems with `pointsinside`.

```r
load("/GD/LMBD/Papers/2017FAFB/bocklab_public/temca2data/geometry_analysis/data/pns.RData")
lhr=as.mesh3d(subset(FAFB13NP.surf, 'LH_R'))
pns.lh=nlapply(pns, subset, function(x) pointsinside(x, lhr))
```
pns.lh includes a lot of stuff outside the LH for some reason – only things that are distant though.

One option, given below would be to use a bounding box check to speed up computation as well as rejecting distant points.

```r
# pointsinside seems to be misbehaving for points a long way away
contains_points <- function(obj, points, ...) UseMethod("contains_points")

contains_points.boundingbox <- function(obj, points,  ...) {
  xyz=xyzmatrix(points)
  xyz[,1] >= obj[1,1] & xyz[,2] >= obj[1,2] & xyz[,3] >= obj[1,3] &
    xyz[,1] <= obj[2,1] & xyz[,2] <= obj[2,2] & xyz[,3] <= obj[2,3]
}

contains_points.mesh3d <- function(obj, points,  ...) {
  xyz=xyzmatrix(points)
  inbb=contains_points(boundingbox(obj), xyz, ...)
  iosurf=pointsinside(xyz[inbb,], surf = obj, ...)
  res=inbb
  res[inbb]=iosurf
  res
}

contains_points.hxsurf<-contains_points.mesh3d
```

**Comment by jefferis** [comment]
Just to note that the example above `pns.lh` works fine after Rvcg commit: https://github.com/zarquon42b/Rvcg/commit/63329d01827c267d4ad07dbfe7b182b30601c5a7

**Comment by jefferis** [comment]
Could also add a method for `ashape3d` objects

**Comment by jefferis** [comment]
```r
contains_points.mesh3d <- function(obj, points,  ...) {
  xyz=xyzmatrix(points)
  inbb=contains_points(boundingbox(obj), xyz, ...)
  iosurf=pointsinside(xyz[inbb,,drop=F], surf = obj, ...)
  res=inbb
  res[inbb]=iosurf
  res
}
```

**Comment by jefferis** [comment]
Initial work on https://github.com/jefferis/nat/tree/feature/improve-pointsinside (see ac15aa99631f319252a6a9200bbf7b6afe1ef7a5)

**Comment by jefferis** [comment]
@fmlove this issue will be of interest

**Comment by jefferis** [comment]
Have to decided to merge to master as is (i.e. 0cc65c09e7e82744b0aa93dcb35dbc1b2ff9f755). This may not be the last word, but it does seem to solve the majority of issues in the wild including the issues faced by @fmlove.

**Comment by jefferis** [comment]
closed by 3362ad87bbda2584fba7b5578ed9a6cdc6bb57af
